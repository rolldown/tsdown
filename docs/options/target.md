# Target

The `target` setting determines which JavaScript and CSS features are downleveled (transformed to older syntax) and which are left intact in the output. This allows you to control the compatibility of your bundled code with specific environments or JavaScript versions.

For example, an arrow function `() => this` will be transformed into an equivalent `function` expression if the target is `es5` or lower.

> [!WARNING] Syntax Downgrade Only
> The `target` option only affects syntax transformations. It does not include runtime polyfills or shims for APIs that may not exist in the target environment. For example, if your code uses `Promise`, it will not be polyfilled for environments that lack native `Promise` support.

## Default Target Behavior

By default, `tsdown` will read the `engines.node` field from your `package.json` and automatically set the target to the minimum compatible Node.js version specified. This ensures your output is compatible with the environments you declare for your package.

For example, if your `package.json` contains:

```json
{
  "engines": {
    "node": ">=18.0.0"
  }
}
```

Then `tsdown` will automatically set the target to `node18.0.0`.

If you want to override this behavior, you can specify the target explicitly using the CLI or configuration file.

## Disabling Target Transformations

You can disable all syntax transformations by setting the target to `false`. This will preserve modern JavaScript and CSS syntax in the output, regardless of the environment specified in your `package.json`.

```json
{
  "target": false
}
```

When `target` is set to `false`:

- No JavaScript syntax downleveling occurs (modern features like optional chaining `?.`, nullish coalescing `??`, etc. are preserved)
- No CSS syntax transformations are applied (modern CSS features like nesting are preserved)
- No runtime helper plugins are loaded
- The output will use the exact syntax from your source code

This is particularly useful when:

- You're targeting modern environments that support the latest JavaScript/CSS features
- You want to handle syntax transformations in a different build step
- You're building a library that will be further processed by the consuming application

> [!NOTE] No Target Resolution
> If you don't specify a `target` and your `package.json` doesn't have an `engines.node` field, `tsdown` will behave as if `target: false` was set, preserving all modern syntax.

## Customizing the Target

You can specify the target using the `--target` option:

```bash
tsdown --target <target>
```

### Supported Targets

- ECMAScript versions: `es5`, `es2015`, `es2020`, `esnext`, etc.
- Browser versions: `chrome100`, `safari18`, `firefox110`, etc.
- Node.js versions: `node20.18`, `node16`, etc.

### Example

```bash
tsdown --target es2020
```

You can also pass an array of targets to ensure compatibility across multiple environments:

```bash
tsdown --target chrome100 --target node20.18
```

# CSS Targeting

`tsdown` can also downlevel CSS features to match your specified browser targets. For example, a CSS nesting `&` selector will be flattened if the target is `chrome108` or lower.

To enable CSS downleveling, you need to manually install [`unplugin-lightningcss`](https://github.com/unplugin/unplugin-lightningcss):

::: code-group

```sh [npm]
npm install -D unplugin-lightningcss
```

```sh [pnpm]
pnpm add -D unplugin-lightningcss
```

```sh [yarn]
yarn add -D unplugin-lightningcss
```

```sh [bun]
bun add -D unplugin-lightningcss
```

:::

Once installed, simply set your browser target (for example, `target: 'chrome100'`) in your configuration or CLI options, and CSS downleveling will be enabled automatically.

For more information on browser targets and CSS compatibility, refer to the [Lightning CSS documentation](https://lightningcss.dev/).
