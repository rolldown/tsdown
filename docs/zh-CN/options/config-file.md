# 配置文件

默认情况下，`tsdown` 会在当前工作目录中查找配置文件，并向上遍历父目录直到找到一个配置文件。它支持以下文件名：

- `tsdown.config.ts`
- `tsdown.config.mts`
- `tsdown.config.cts`
- `tsdown.config.js`
- `tsdown.config.mjs`
- `tsdown.config.cjs`
- `tsdown.config.json`
- `tsdown.config`

此外，您还可以直接在 `package.json` 文件的 `tsdown` 字段中定义配置。

## 编写配置文件

配置文件允许您以集中且可复用的方式定义和自定义构建设置。以下是一个简单的 `tsdown` 配置文件示例：

```ts [tsdown.config.ts]
import { defineConfig } from 'tsdown'

export default defineConfig({
  entry: 'src/index.ts',
})
```

### 构建多个输出

`tsdown` 还支持从配置文件返回一个**配置数组**。这允许您在一次运行中使用不同的设置构建多个输出。例如：

```ts [tsdown.config.ts]
import { defineConfig } from 'tsdown'

export default defineConfig([
  {
    entry: 'src/entry1.ts',
    platform: 'node',
  },
  {
    entry: 'src/entry2.ts',
    platform: 'browser',
  },
])
```

## 指定自定义配置文件

如果您的配置文件位于其他位置或具有不同的名称，可以使用 `--config`（或 `-c`）选项指定其路径：

```bash
tsdown --config ./path/to/config
```

## 禁用配置文件 {#disable-config-file}

如果您希望完全禁用加载配置文件，可以使用 `--no-config` 选项：

```bash
tsdown --no-config
```

这在您希望仅依赖命令行选项或默认设置时非常有用。

## 配置加载器

`tsdown` 支持多种配置加载器，以适配不同的文件格式。您可以通过 `--config-loader` 选项选择配置加载器。可用的加载器包括：

- `auto`（默认）：如果运行时支持，则使用原生方式加载 TypeScript，否则回退到 `unconfig`。
- `native`：通过原生运行时支持加载 TypeScript 配置文件。需要兼容的环境，如最新版 Node.js、Deno 或 Bun。
- `unconfig`：使用 `unconfig` 库加载配置文件，确保在不同运行时下具有广泛兼容性。
- `unrun` _(实验性)_: 使用 [`unrun`](https://gugustinette.github.io/unrun/) 库加载配置文件，兼容性与 `unconfig` 类似，但性能更高。要使用此加载器，请先安装 `unrun` 包。

> [!TIP]
> Node.js 原生不支持在不指定文件扩展名的情况下导入 TypeScript 文件。如果您在 Node.js 环境下希望加载不带 `.ts` 扩展名的 TypeScript 配置文件，建议使用 `unconfig` 或 `unrun` 加载器以获得更好的兼容性。

## 扩展 Vite 或 Vitest 配置（实验性功能）{#extending-vite-or-vitest-config-experimental}

`tsdown` 提供了一个**实验性**功能，允许您扩展现有的 Vite 或 Vitest 配置文件。通过此功能，您可以复用特定的配置选项（如 `resolve` 和 `plugins`），同时忽略与 `tsdown` 无关的其他选项。

要启用此功能，请使用 `--from-vite` 选项：

```bash
tsdown --from-vite        # 加载 vite.config.*
tsdown --from-vite vitest # 加载 vitest.config.*
```

> [!WARNING]
> 此功能为 **实验性功能**，可能并不支持所有 Vite 或 Vitest 的配置选项。仅特定选项（如 `resolve` 和 `plugins`）会被重用。请谨慎使用，并在您的项目中充分测试。

> [!TIP]
> 如果您的项目已经使用了 Vite 或 Vitest，扩展其配置可以节省时间和精力，让您在现有设置的基础上构建，而无需重复配置。

## 参考

有关可用配置选项的完整列表，请参阅 [配置选项参考](../reference/api/Interface.Options.md)。其中包括所有支持字段及其用法的详细说明。
