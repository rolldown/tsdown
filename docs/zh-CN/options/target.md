# 构建目标（Target）

`target` 设置决定了哪些 JavaScript 和 CSS 特性会被降级（转换为旧语法），哪些会在输出中保持原样。这使您可以控制打包代码与特定环境或 JavaScript 版本的兼容性。

例如，如果目标是 `es5` 或更低版本，箭头函数 `() => this` 将被转换为等效的 `function` 表达式。

> [!WARNING] 仅限语法降级
> `target` 选项仅影响语法转换。它不会为目标环境中可能不存在的 API 提供运行时 polyfill 或 shim。例如，如果您的代码使用了 `Promise`，但目标环境不支持原生 `Promise`，则不会自动添加 polyfill。

## 默认目标行为

默认情况下，`tsdown` 会读取您的 `package.json` 中的 `engines.node` 字段，并自动将目标设置为所声明的最低兼容 Node.js 版本。这可以确保您的输出与您为包声明的运行环境兼容。

例如，如果您的 `package.json` 包含：

```json
{
  "engines": {
    "node": ">=18.0.0"
  }
}
```

那么 `tsdown` 会自动将目标设置为 `node18.0.0`。

如果您希望覆盖此行为，可以通过 CLI 或配置文件显式指定目标。

## 禁用目标转换

您可以通过将目标设置为 `false` 来禁用所有语法转换。这将在输出中保留现代 JavaScript 和 CSS 语法，无论您的 `package.json` 中指定的环境如何。

```json
{
  "target": false
}
```

当 `target` 设置为 `false` 时：
- 不会进行 JavaScript 语法降级（现代特性如可选链 `?.`、空值合并 `??` 等会被保留）
- 不会应用 CSS 语法转换（现代 CSS 特性如嵌套会被保留）
- 不会加载运行时辅助插件
- 输出将使用源代码中的确切语法

这在以下情况下特别有用：
- 您的目标是支持最新 JavaScript/CSS 特性的现代环境
- 您希望在不同的构建步骤中处理语法转换
- 您正在构建一个将由使用应用程序进一步处理的库

> [!NOTE] 无目标解析
> 如果您没有指定 `target` 且您的 `package.json` 没有 `engines.node` 字段，`tsdown` 将表现得就像设置了 `target: false` 一样，保留所有现代语法。

## 自定义目标

您可以使用 `--target` 选项指定目标：

```bash
tsdown --target <target>
```

### 支持的目标

- ECMAScript 版本：`es5`、`es2015`、`es2020`、`esnext` 等
- 浏览器版本：`chrome100`、`safari18`、`firefox110` 等
- Node.js 版本：`node20.18`、`node16` 等

### 示例

```bash
tsdown --target es2020
```

您还可以传递多个目标，以确保兼容多个环境：

```bash
tsdown --target chrome100 --target node20.18
```

# CSS 目标

`tsdown` 也可以将 CSS 特性降级以匹配您指定的浏览器目标。例如，如果目标是 `chrome108` 或更低版本，CSS 嵌套的 `&` 选择器将被展开为平铺结构。

要启用 CSS 降级，您需要手动安装 [`unplugin-lightningcss`](https://github.com/unplugin/unplugin-lightningcss)：

::: code-group

```sh [npm]
npm install -D unplugin-lightningcss
```

```sh [pnpm]
pnpm add -D unplugin-lightningcss
```

```sh [yarn]
yarn add -D unplugin-lightningcss
```

```sh [bun]
bun add -D unplugin-lightningcss
```

:::

安装后，只需在配置或 CLI 选项中设置您的浏览器目标（例如 `target: 'chrome100'`），CSS 降级将会自动启用。

有关浏览器目标和 CSS 兼容性的更多信息，请参阅 [Lightning CSS 文档](https://lightningcss.dev/)。
