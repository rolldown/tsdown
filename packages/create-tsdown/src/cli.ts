import process from 'node:process'
import { log } from '@clack/prompts'
import { cac } from 'cac'
import { version } from '../package.json'
import { create, type Options } from './index'

const cli = cac('create-tsdown')
cli.help().version(version)

cli
  .command('[path]', 'Create a tsdown project', {
    ignoreOptionDefaultValue: true,
    allowUnknownOptions: true,
  })
  .option(
    '-t, --template <template>',
    'Available templates: default, minimal, vue, react, solid',
    { default: 'default' },
  )
  .action((path: string | undefined, options: Options) => create(path, options))

export async function runCLI(): Promise<void> {
  cli.parse(process.argv, { run: false })

  try {
    await cli.runMatchedCommand()
  } catch (error) {
    log.error(String(error))
    process.exit(1)
  }
}
