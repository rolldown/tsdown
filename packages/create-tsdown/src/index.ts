import process from 'node:process'
import { styleText } from 'node:util'
import {
  cancel,
  intro,
  isCancel,
  outro,
  select,
  spinner,
  text,
} from '@clack/prompts'
import { downloadTemplate } from 'giget'
import { getUserAgent } from 'package-manager-detector'

export interface Options {
  template?: 'default' | 'minimal' | 'vue' | 'react' | 'solid'
  path?: string
}

export type ResolvedOptions = Required<Options>

/**
 * Create a tsdown project.
 */
export async function create(
  path: string | undefined,
  options: Options,
): Promise<void> {
  intro(`Creating a tsdown project...`)

  const resolved = await resolveOptions({ ...options, path })

  const s = spinner()
  s.start('Cloning the template...')

  await downloadTemplate(`gh:sxzz/tsdown-templates/${resolved.template}`, {
    dir: resolved.path,
  })

  s.stop('Template cloned')

  const pm = getUserAgent() || 'npm'
  outro(
    `Done! Now run:\n` +
      `  ${styleText('green', `cd ${resolved.path}`)}\n` +
      `  ${styleText('green', `${pm} install`)}\n` +
      `  ${styleText('green', `${pm} run build`)}\n\n` +
      `For more information, visit: ${styleText('underline', `https://tsdown.dev/`)}`,
  )
}

/**
 * Resolve the user options and configs
 * @param options The user options
 * @returns The resolved options
 */
export async function resolveOptions(
  options: Options,
): Promise<ResolvedOptions> {
  let path: Options['path'] | symbol = options.path

  if (!path) {
    const defaultPath = './my-tsdown-package'
    path =
      (await text({
        message: 'What is the name of your package?',
        placeholder: defaultPath,
      })) || defaultPath
    if (isCancel(path)) {
      cancel('Operation cancelled.')
      process.exit(1)
    }
  }

  let template: Options['template'] | symbol = options.template
  if (template) {
    if (!['default', 'minimal', 'vue', 'react', 'solid'].includes(template)) {
      throw new Error(
        `Invalid template "${template}". Available templates: default, vue, react, solid`,
      )
    }
  } else {
    template = await select({
      message: 'Which template do you want to use?',
      options: [
        { value: 'default', label: 'Default' },
        { value: 'minimal', label: 'Minimal' },
        { value: 'vue', label: 'Vue' },
        { value: 'react', label: 'React' },
        { value: 'solid', label: 'Solid' },
      ],
      initialValue: 'default',
    })

    if (isCancel(template)) {
      cancel('Operation cancelled.')
      process.exit(1)
    }
  }

  return {
    path,
    template,
  } satisfies ResolvedOptions
}
