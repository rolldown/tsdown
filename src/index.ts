import path from 'node:path'
import process from 'node:process'
import { green } from 'ansis'
import Debug from 'debug'
import {
  build as rolldownBuild,
  type BuildOptions,
  type OutputOptions,
  type RolldownPluginOption,
} from 'rolldown'
import { transformPlugin } from 'rolldown/experimental'
import { exec } from 'tinyexec'
import { cleanOutDir } from './features/clean'
import { copy } from './features/copy'
import { ExternalPlugin } from './features/external'
import { createHooks } from './features/hooks'
import { LightningCSSPlugin } from './features/lightningcss'
import { NodeProtocolPlugin } from './features/node-protocol'
import { resolveChunkFilename } from './features/output'
import { publint } from './features/publint'
import { ReportPlugin } from './features/report'
import { getShimsInject } from './features/shims'
import { shortcuts } from './features/shortcuts'
import { watchBuild } from './features/watch'
import {
  mergeUserOptions,
  resolveOptions,
  type NormalizedFormat,
  type Options,
  type ResolvedOptions,
} from './options'
import { ShebangPlugin } from './plugins'
import { logger, setSilent } from './utils/logger'
import { prettyFormat } from './utils/package'
import type { Options as DtsOptions } from 'rolldown-plugin-dts'

const debug = Debug('tsdown:main')

/**
 * Build with tsdown.
 */
export async function build(userOptions: Options = {}): Promise<void> {
  if (typeof userOptions.silent === 'boolean') {
    setSilent(userOptions.silent)
  }

  debug('Loading config')
  const { configs, file: configFile } = await resolveOptions(userOptions)
  if (configFile) {
    debug('Loaded config:', configFile)
    configs.forEach((config) => {
      debug('using resolved config: %O', config)
    })
  } else {
    debug('No config file found')
  }

  let cleanPromise: Promise<void> | undefined
  const clean = () => {
    if (cleanPromise) return cleanPromise
    return (cleanPromise = cleanOutDir(configs))
  }

  const rebuilds = await Promise.all(
    configs.map((options) => buildSingle(options, clean)),
  )
  const cleanCbs: (() => Promise<void>)[] = []

  for (const [i, config] of configs.entries()) {
    const rebuild = rebuilds[i]
    if (!rebuild) continue

    const watcher = await watchBuild(config, configFile, rebuild, restart)
    cleanCbs.push(() => watcher.close())
  }

  if (cleanCbs.length) {
    shortcuts(restart)
  }

  async function restart() {
    for (const clean of cleanCbs) {
      await clean()
    }
    build(userOptions)
  }
}

export const pkgRoot: string = path.resolve(import.meta.dirname, '..')

/**
 * Build a single configuration, without watch and shortcuts features.
 *
 * Internal API, not for public use
 *
 * @private
 * @param config Resolved options
 */
export async function buildSingle(
  config: ResolvedOptions,
  clean: () => Promise<void>,
): Promise<(() => Promise<void>) | undefined> {
  const { format: formats, dts, watch, onSuccess } = config
  let onSuccessCleanup: (() => any) | undefined

  const { hooks, context } = await createHooks(config)

  await rebuild(true)
  if (watch) {
    return () => rebuild()
  }

  async function rebuild(first?: boolean) {
    const startTime = performance.now()

    await hooks.callHook('build:prepare', context)
    onSuccessCleanup?.()

    await clean()

    let hasErrors = false
    await Promise.all(
      formats.map(async (format) => {
        try {
          const formatLabel = prettyFormat(format)
          logger.info(formatLabel, 'Build start')

          const buildOptions = await getBuildOptions(config, format)
          await hooks.callHook('build:before', {
            ...context,
            buildOptions,
          })
          await rolldownBuild(buildOptions)
          if (format === 'cjs' && dts) {
            await rolldownBuild(await getBuildOptions(config, format, true))
          }
        } catch (error) {
          if (watch) {
            logger.error(error)
            hasErrors = true
            return
          }
          throw error
        }
      }),
    )

    if (hasErrors) {
      return
    }

    await publint(config)
    await copy(config)

    await hooks.callHook('build:done', context)

    logger.success(
      `${first ? 'Build' : 'Rebuild'} complete in ${green(`${Math.round(performance.now() - startTime)}ms`)}`,
    )

    if (typeof onSuccess === 'string') {
      const p = exec(onSuccess, [], {
        nodeOptions: { shell: true, stdio: 'inherit' },
      })
      p.then(({ exitCode }) => {
        if (exitCode) {
          process.exitCode = exitCode
        }
      })
      onSuccessCleanup = () => p.kill('SIGTERM')
    } else {
      await onSuccess?.(config)
    }
  }
}

async function getBuildOptions(
  config: ResolvedOptions,
  format: NormalizedFormat,
  cjsDts?: boolean,
): Promise<BuildOptions> {
  const {
    entry,
    external,
    plugins: userPlugins,
    outDir,
    platform,
    alias,
    treeshake,
    sourcemap,
    dts,
    minify,
    unused,
    target,
    define,
    shims,
    tsconfig,
    cwd,
    report,
    env,
    removeNodeProtocol,
  } = config

  const plugins: RolldownPluginOption = []
  if (config.pkg || config.skipNodeModulesBundle) {
    plugins.push(ExternalPlugin(config))
  }

  if (dts) {
    const { dts: dtsPlugin } = await import('rolldown-plugin-dts')
    const options: DtsOptions = { tsconfig, ...dts }

    if (format === 'es') {
      plugins.push(dtsPlugin(options))
    } else if (cjsDts) {
      plugins.push(dtsPlugin({ ...options, emitDtsOnly: true }))
    }
  }
  if (!cjsDts) {
    if (unused) {
      const { Unused } = await import('unplugin-unused')
      plugins.push(Unused.rolldown(unused === true ? {} : unused))
    }
    if (target) {
      plugins.push(
        transformPlugin({
          include: /\.[cm]?[jt]sx?$/,
          exclude: /\.d\.[cm]?ts$/,
          transformOptions: {
            target,
          },
        }),
      )
    }
    plugins.push(ShebangPlugin(cwd))
  }

  if (report && logger.level >= 3) {
    plugins.push(ReportPlugin(report, cwd, cjsDts))
  }

  if (target) {
    plugins.push(
      // Use Lightning CSS to handle CSS input. This is a temporary solution
      // until Rolldown supports CSS syntax lowering natively.
      LightningCSSPlugin({ target }),
    )
  }

  plugins.push(userPlugins)

  if (removeNodeProtocol) {
    plugins.push(NodeProtocolPlugin())
  }

  const inputOptions = await mergeUserOptions(
    {
      input: entry,
      cwd,
      external,
      resolve: {
        alias,
        tsconfigFilename: tsconfig || undefined,
      },
      treeshake,
      platform,
      define: {
        ...define,
        ...Object.keys(env).reduce((acc, key) => {
          const value = JSON.stringify(env[key])
          acc[`process.env.${key}`] = value
          acc[`import.meta.env.${key}`] = value
          return acc
        }, Object.create(null)),
      },
      plugins,
      inject: {
        ...(shims && !cjsDts && getShimsInject(format, platform)),
      },
    },
    config.inputOptions,
    [format],
  )

  const [entryFileNames, chunkFileNames] = resolveChunkFilename(
    config,
    inputOptions,
    format,
  )
  const outputOptions: OutputOptions = await mergeUserOptions(
    {
      format: cjsDts ? 'es' : format,
      name: config.globalName,
      sourcemap,
      dir: outDir,
      minify: minify as any,
      entryFileNames,
      chunkFileNames,
    },
    config.outputOptions,
    [format],
  )

  return {
    ...inputOptions,
    output: outputOptions,
  }
}

export { defineConfig } from './config'
export type { Options, UserConfig } from './options'
export type { BuildContext, TsdownHooks } from './features/hooks'
export { logger }
