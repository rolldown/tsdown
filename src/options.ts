import { stat } from 'node:fs/promises'
import path from 'node:path'
import process from 'node:process'
import { loadConfig } from 'unconfig'
import { resolveEntry } from './features/entry'
import { toArray } from './utils/general'
import { logger } from './utils/logger'
import type { External } from './features/external'
import type { MarkPartial, MaybePromise, Overwrite } from './utils/types'
import type { Stats } from 'node:fs'
import type { InputOptions, OutputOptions } from 'rolldown'
import type { Options as IsolatedDeclOptions } from 'unplugin-isolated-decl'

export type Format = 'es' | 'esm' | 'module' | 'cjs' | 'commonjs'
export type Sourcemap = boolean | 'inline' | 'hidden'

export interface Options {
  entry?: InputOptions['input']
  format?: Format | Format[]
  plugins?: InputOptions['plugins']
  external?: External
  outDir?: string
  clean?: boolean | string[]
  silent?: boolean
  config?: boolean | string
  sourcemap?: Sourcemap
  alias?: Record<string, string>
  /** @default true */
  treeshake?: boolean
  /** @default false */
  minify?: boolean
  /** @default 'node' */
  platform?: 'node' | 'neutral'
  /**
   * Enable dts generation with `isolatedDeclarations` (experimental)
   */
  dts?: boolean | IsolatedDeclOptions
  watch?: boolean | string | string[]
  inputOptions?: InputOptions
  outputOptions?:
    | OutputOptions
    | ((
        options: OutputOptions,
        format: Format,
      ) => MaybePromise<OutputOptions | void | null>)
}

export type OptionsWithoutConfig = Omit<Options, 'config'>

export type ResolvedOptions = Omit<
  Overwrite<
    MarkPartial<
      Options,
      'inputOptions' | 'outputOptions' | 'minify' | 'alias' | 'external'
    >,
    { format: Format[]; clean: string[] | false }
  >,
  'config'
>

export async function normalizeOptions(
  options: Options,
): Promise<ResolvedOptions> {
  options = {
    ...(await loadConfigFile(options)),
    ...options,
  }

  let {
    entry,
    format = ['es'],
    plugins = [],
    external,
    clean = false,
    silent = false,
    treeshake = true,
    platform = 'node',
    outDir = 'dist',
    sourcemap = false,
    dts = false,
    minify,
    alias,
    watch = false,
    inputOptions,
    outputOptions,
  } = options

  entry = await resolveEntry(entry)
  format = toArray(format, 'es')
  if (clean === true) clean = []

  return {
    entry,
    plugins,
    external,
    format,
    outDir,
    clean,
    silent,
    alias,
    treeshake,
    platform,
    sourcemap,
    dts,
    minify,
    watch,
    inputOptions,
    outputOptions,
  }
}

async function loadConfigFile(options: Options): Promise<Options> {
  let { config: filePath } = options
  if (filePath === false) return {}

  let cwd = process.cwd()
  let overrideConfig = false
  let stats: Stats | null

  if (
    typeof filePath === 'string' &&
    (stats = await stat(filePath).catch(() => null))
  ) {
    const resolved = path.resolve(filePath)
    if (stats.isFile()) {
      overrideConfig = true
      filePath = resolved
      cwd = path.dirname(filePath)
    } else {
      cwd = resolved
    }
  }

  const { config, sources } = await loadConfig<Options>({
    sources: overrideConfig
      ? [{ files: filePath as string, extensions: [] }]
      : [
          {
            files: 'tsdown.config',
            extensions: ['ts', 'mts', 'cts', 'js', 'mjs', 'cjs', 'json', ''],
          },
          {
            files: 'package.json',
            extensions: [],
            rewrite: (config: any) => config?.tsdown,
          },
        ],
    cwd,
    defaults: {},
  })

  if (sources.length > 0) {
    logger.info(`Using tsdown config: ${sources.join(', ')}`)
  }

  return config
}
