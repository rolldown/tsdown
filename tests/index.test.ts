import { beforeEach, expect, test, vi } from 'vitest'
import { resolveOptions } from '../src/options'
import { fsRemove } from '../src/utils/fs'
import { getTestDir, testBuild, writeFixtures } from './utils'

beforeEach(async (context) => {
  const dir = getTestDir(context.task)
  await fsRemove(dir)
})

test('basic', async (context) => {
  const content = `console.log("Hello, world!")`
  const { snapshot } = await testBuild({
    context,
    files: {
      'index.ts': content,
    },
  })
  expect(snapshot).contain(content)
})

{
  const files = {
    'index.ts': "export { foo } from './foo'",
    'foo.ts': 'export const foo = 1',
  }
  test('esm import', async (context) => {
    await testBuild({ context, files })
  })

  test('cjs import', async (context) => {
    await testBuild({
      context,
      files,
      options: {
        format: 'cjs',
      },
    })
  })
}

test('esm shims', async (context) => {
  await testBuild({
    context,
    files: { 'index.ts': 'export default [__dirname, __filename]' },
    options: { shims: true },
  })
})

test('cjs shims', async (context) => {
  const { snapshot } = await testBuild({
    context,
    files: {
      'index.ts': `
        export const url = import.meta.url
        export const filename = import.meta.filename
        export const dirname = import.meta.dirname`,
    },
    options: {
      shims: true,
      format: 'cjs',
    },
  })
  expect(snapshot).contain('require("url").pathToFileURL(__filename).href')
  expect(snapshot).contain('__filename')
  expect(snapshot).contain('__dirname')
})

test('entry structure', async (context) => {
  const files = {
    'src/index.ts': '',
    'src/utils/index.ts': '',
  }
  await testBuild({
    context,
    files,
    options: {
      entry: Object.keys(files),
    },
  })
})

test('bundle dts', async (context) => {
  const files = {
    'src/index.ts': `
      export { str } from './utils/types';
      export { shared } from './utils/shared';
      `,
    'src/utils/types.ts': 'export let str = "hello"',
    'src/utils/shared.ts': 'export let shared = 10',
  }
  await testBuild({
    context,
    files,
    options: {
      entry: ['src/index.ts'],
      dts: true,
    },
  })
})

test('cjs interop', async (context) => {
  const files = {
    'index.ts': `
    export default {}
    export type Foo = string
    `,
  }
  await testBuild({
    context,
    files,
    options: {
      format: ['esm', 'cjs'],
    },
  })
})

test('fixed extension', async (context) => {
  const files = {
    'index.ts': `export default 10`,
  }
  await testBuild({
    context,
    files,
    options: {
      format: ['esm', 'cjs'],
      fixedExtension: true,
      dts: true,
    },
  })
})

test('custom extension', async (context) => {
  const files = {
    'index.ts': `export default 10`,
  }
  const { outputFiles } = await testBuild({
    context,
    files,
    options: {
      dts: true,
      outExtensions: () => ({ js: '.some.mjs', dts: '.some.d.mts' }),
    },
  })
  expect(outputFiles).toMatchInlineSnapshot(`
    [
      "index.d.some.d.mts",
      "index.some.mjs",
    ]
  `)
})

test('noExternal', async (context) => {
  const files = {
    'index.ts': `export * from 'cac'`,
  }
  await testBuild({
    context,
    files,
    options: {
      noExternal: ['cac'],
      plugins: [
        {
          name: 'remove-code',
          load(id) {
            if (id.replaceAll('\\', '/').includes('/node_modules/cac')) {
              return 'export const cac = "[CAC CODE]"'
            }
          },
        },
      ],
    },
  })
})

test('fromVite', async (context) => {
  const files = {
    'index.ts': `export default 10`,
    'tsdown.config.ts': `
    import { resolve } from 'node:path'
    export default {
      entry: resolve(import.meta.dirname, 'index.ts').replaceAll('\\\\', '/'),
      fromVite: true,
    }`,
    'vite.config.ts': `
    export default {
      resolve: { alias: { '~': '/' } },
      plugins: [{ name: 'expected' }],
    }
    `,
  }
  const { testDir } = await writeFixtures(context, files)
  const options = await resolveOptions({
    config: testDir,
  })
  expect(options.configs).toMatchObject([
    {
      fromVite: true,
      alias: {
        '~': '/',
      },
      plugins: [
        [
          {
            name: 'expected',
          },
        ],
        [],
      ],
    },
  ])
})

test('resolve dependency for dts', async (context) => {
  const files = {
    'index.ts': `export type { GlobOptions } from 'tinyglobby'
    export type * from 'unconfig'`,
  }
  const { snapshot } = await testBuild({
    context,
    files,
    options: {
      dts: { resolve: ['tinyglobby'] },
    },
  })
  expect(snapshot).contain(`export * from "unconfig"`)
})

test('resolve paths in tsconfig', async (context) => {
  const files = {
    'index.ts': `export * from '@/mod'`,
    'mod.ts': `export const mod = 42`,
    '../tsconfig.build.json': JSON.stringify({
      compilerOptions: {
        paths: { '@/*': ['./resolve-paths-in-tsconfig/*'] },
      },
    }),
  }
  await testBuild({
    context,
    files,
    options: {
      dts: { isolatedDeclarations: true },
      tsconfig: 'tsconfig.build.json',
    },
  })
})

test('hooks', async (context) => {
  const fn = vi.fn()
  const files = {
    'index.ts': `export default 10`,
  }
  await testBuild({
    context,
    files,
    options: {
      hooks: {
        'build:prepare': fn,
        'build:before': fn,
        'build:done': fn,
      },
    },
  })
  expect(fn).toBeCalledTimes(3)
})

test('env flag', async (context) => {
  const files = {
    'index.ts': `export const env = process.env.NODE_ENV
    export const meta = import.meta.env.NODE_ENV
    export const custom = import.meta.env.CUSTOM
    export const debug = import.meta.env.DEBUG
    `,
  }
  const { snapshot } = await testBuild({
    context,
    files,
    options: {
      env: {
        NODE_ENV: 'production',
        CUSTOM: 'tsdown',
        DEBUG: true,
      },
    },
  })
  expect(snapshot).contains('const env = "production"')
  expect(snapshot).contains('const meta = "production"')
  expect(snapshot).contains('const custom = "tsdown"')
  expect(snapshot).contains('const debug = true')
})

test('minify', async (context) => {
  const files = { 'index.ts': `export const foo = true` }
  const { snapshot } = await testBuild({
    context,
    files,
    options: {
      minify: {
        mangle: true,
        compress: true,
        removeWhitespace: false,
      },
    },
  })
  expect(snapshot).contains('!0')
  expect(snapshot).not.contains('true')
  expect(snapshot).not.contains('const foo')
})

test('iife and umd', async (context) => {
  const files = { 'index.ts': `export const foo = true` }
  const { outputFiles } = await testBuild({
    context,
    files,
    options: {
      format: ['iife', 'umd'],
      globalName: 'Lib',
    },
  })
  expect(outputFiles).toMatchInlineSnapshot(`
    [
      "index.iife.js",
      "index.umd.js",
    ]
  `)
})

test('remove node protocol', async (context) => {
  const files = {
    'index.ts': `
    import fs from 'node:fs'
    import { join } from 'node:path'
    const promise = import('node:fs/promises')

    export { fs, join, promise }
    `,
  }
  const { snapshot } = await testBuild({
    context,
    files,
    options: {
      removeNodeProtocol: true,
    },
  })
  expect(snapshot).not.contains('node:')
})
