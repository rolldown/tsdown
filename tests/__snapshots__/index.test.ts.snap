// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
"//#region index.ts
console.log("Hello, world!");

//#endregion"
`;

exports[`cjs import 1`] = `
""use strict";

//#region foo.ts
const foo = 1;

//#endregion
Object.defineProperty(exports, 'foo', {
  enumerable: true,
  get: function () {
    return foo;
  }
});"
`;

exports[`esm import 1`] = `
"//#region foo.ts
const foo = 1;

//#endregion
export { foo };"
`;
